File: src/omnia/data/structure/observable/writable/WritableObservableListImpl.java
package omnia.data.structure.observable.writable;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.processors.FlowableProcessor;
import io.reactivex.processors.PublishProcessor;
import java.util.Iterator;
import java.util.OptionalInt;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Stream;
import omnia.data.structure.Collection;
import omnia.data.structure.IntRange;
import omnia.data.structure.List;
import omnia.data.structure.immutable.ImmutableList;
import omnia.data.structure.observable.ObservableList;

final class WritableObservableListImpl<E> implements WritableObservableList<E> {

  private volatile ImmutableList<E> currentState = ImmutableList.empty();
  private final FlowableProcessor<MutationEvent> mutationEventProcessor =
      PublishProcessor.<MutationEvent>create().toSerialized();
  private final ObservableChannels observableChannels = new ObservableChannels();

  @Override
  public void insertAt(int index, E item) {
    IntRange range = IntRange.startingAt(index).withLength(1);
    mutateState(
        state -> true,
        state -> state.toBuilder().insertAt(index, item).build(),
        (previousState, currentState) -> generateInsertAtMutations(currentState, range));
  }

  @Override
  public E removeAt(int index) {
    IntRange range = IntRange.startingAt(index).withLength(1);
    E value;
    synchronized (this) {
      value = getState().itemAt(index);
      mutateState(
          state -> true,
          state -> state.toBuilder().removeAt(index).build(),
          (previousState, currentState) -> generateRemoveAtMutations(previousState, range));
    }
    return value;
  }

  @Override
  public E replaceAt(int index, E item) {
    IntRange range = IntRange.startingAt(index).withLength(1);
    E value;
    synchronized (this) {
      value = getState().itemAt(index);
      mutateState(
          state -> true,
          state -> state.toBuilder().replaceAt(index, item).build(),
          (previousState, currentState) ->
              generateReplaceAtMutations(previousState, currentState, range));
    }
    return value;
  }

  @Override
  public void add(E item) {
    mutateState(
        state -> true,
        state -> state.toBuilder().add(item).build(),
        (previousState, currentState) -> generateInsertAtMutations(currentState, IntRange.startingAt(currentState.count() - 1).withLength(1)));
  }

  @Override
  public void addAll(Collection<? extends E> elements) {
    mutateState(
        state -> true,
        state -> state.toBuilder().addAll(elements).build(),
        (previousState, currentState) -> generateInsertAtMutations(
            currentState, IntRange.startingAt(previousState.count()).endingAt(currentState.count())));
  }

  @Override
  public boolean removeUnknownTyped(Object item) {
    synchronized(this) {
      OptionalInt index = getState().indexOf(item);
      return mutateState(
          state -> index.isPresent(),
          state -> state.toBuilder().removeAt(index.orElseThrow()).build(),
          (previousState, currentState) -> generateRemoveAtMutations(
              previousState, IntRange.startingAt(index.orElseThrow()).withLength(1)));
    }
  }

  @Override
  public void clear() {
    synchronized(this) {
      mutateState(
          ImmutableList::isPopulated,
          state -> ImmutableList.empty(),
          (previousState, currentState) ->
              generateRemoveAtMutations(
                  previousState, IntRange.startingAt(0).endingAt(previousState.count())));
    }
  }

  private boolean mutateState(
      Predicate<ImmutableList<E>> shouldMutate,
      Function<ImmutableList<E>, ImmutableList<E>> mutator,
      BiFunction<ImmutableList<E>, ImmutableList<E>, List<ListOperation<E>>> mutationsGenerator) {
    synchronized(this) {
      ImmutableList<E> previousState = currentState;
      if (!shouldMutate.test(previousState)) {
        return false;
      }
      ImmutableList<E> newState = mutator.apply(previousState);
      currentState = newState;
      mutationEventProcessor.onNext(
          new MutationEvent(newState, mutationsGenerator.apply(previousState, newState)));
      return true;
    }
  }

  private static <E> ImmutableList<ListOperation<E>> generateInsertAtMutations(
      ImmutableList<E> state, IntRange range) {
    ImmutableList.Builder<ListOperation<E>> builder = ImmutableList.builder();

    // Move must be done before insert to make way for new items
    if (range.end() < state.count()) {
      int numItemsMoved = state.count() - range.end();
      IntRange moveFromRange = IntRange.startingAt(range.start()).withLength(numItemsMoved);
      IntRange moveToRange = IntRange.startingAt(range.end()).withLength(numItemsMoved);
      builder.add(new MoveInList<>(state.getSublist(moveToRange), moveFromRange, moveToRange));
    }

    builder.add(new AddToList<>(state.getSublist(range), range));

    return builder.build();
  }

  private static <E> ImmutableList<ListOperation<E>> generateRemoveAtMutations(
      ImmutableList<E> previousState, IntRange range) {
    ImmutableList.Builder<ListOperation<E>> builder = ImmutableList.builder();

    /// Removal must be done BEFORE move to make space for moved items
    builder.add(new RemoveFromList<>(previousState.getSublist(range), range));

    if (range.end() < previousState.count()) {
      int numItemsMoved = previousState.count() - range.end();
      IntRange moveFromRange = IntRange.startingAt(range.end()).withLength(numItemsMoved);
      IntRange moveToRange = IntRange.startingAt(range.start()).withLength(numItemsMoved);
      builder.add(
          new MoveInList<>(previousState.getSublist(moveFromRange), moveFromRange, moveToRange));
    }

    return builder.build();
  }

  private static <E> ImmutableList<ListOperation<E>> generateReplaceAtMutations(
      ImmutableList<E> previousState, ImmutableList<E> currentState, IntRange range) {
    return ImmutableList.of(
        new ReplaceInList<>(
            previousState.getSublist(range), currentState.getSublist(range), range));
  }

  @Override
  public Iterator<E> iterator() {
    return getState().iterator();
  }

  @Override
  public boolean containsUnknownTyped(Object element) {
    return getState().containsUnknownTyped(element);
  }

  @Override
  public int count() {
    return getState().count();
  }

  @Override
  public boolean isPopulated() {
    return getState().isPopulated();
  }

  @Override
  public E itemAt(int index) {
    return getState().itemAt(index);
  }

  @Override
  public OptionalInt indexOf(Object item) {
    return getState().indexOf(item);
  }

  @Override
  public Stream<E> stream() {
    return getState().stream();
  }

  @Override
  public ObservableList<E> toReadOnly() {
    return new ObservableList<>() {
      @Override
      public ObservableChannels<E> observe() {
        return WritableObservableListImpl.this.observe();
      }

      @Override
      public Iterator<E> iterator() {
        return WritableObservableListImpl.this.iterator();
      }

      @Override
      public boolean containsUnknownTyped(Object element) {
        return WritableObservableListImpl.this.containsUnknownTyped(element);
      }

      @Override
      public int count() {
        return WritableObservableListImpl.this.count();
      }

      @Override
      public E itemAt(int index) {
        return WritableObservableListImpl.this.itemAt(index);
      }

      @Override
      public OptionalInt indexOf(Object item) {
        return WritableObservableListImpl.this.indexOf(item);
      }

      @Override
      public Stream<E> stream() {
        return WritableObservableListImpl.this.stream();
      }
    };
  }

  @Override
  public ObservableChannels observe() {
    return observableChannels;
  }

  private ImmutableList<E> getState() {
    synchronized(this) {
      return currentState;
    }
  }

  private static final class AddToList<E> implements ObservableList.AddToList<E> {
    private final List<E> items;
    private final IntRange indices;

    private AddToList(List<E> items, IntRange indices) {
      this.items = items;
      this.indices = indices;
    }

    @Override
    public List<E> items() {
      return items;
    }

    @Override
    public IntRange indices() {
      return indices;
    }
  }

  private static final class MoveInList<E> implements ObservableList.MoveInList<E> {
    private final List<E> items;
    private final IntRange previousIndices;
    private final IntRange currentIndices;

    private MoveInList(List<E> items, IntRange previousIndices, IntRange currentIndices) {
      this.items = items;
      this.previousIndices = previousIndices;
      this.currentIndices = currentIndices;
    }

    @Override
    public List<E> items() {
      return items;
    }

    @Override
    public IntRange previousIndices() {
      return previousIndices;
    }

    @Override
    public IntRange currentIndices() {
      return currentIndices;
    }
  }

  private static final class RemoveFromList<E> implements ObservableList.RemoveFromList<E> {
    private final List<E> items;
    private final IntRange indices;

    private RemoveFromList(List<E> items, IntRange indices) {
      this.items = items;
      this.indices = indices;
    }

    @Override
    public List<E> items() {
      return items;
    }

    @Override
    public IntRange indices() {
      return indices;
    }
  }

  private static final class ReplaceInList<E> implements ObservableList.ReplaceInList<E> {
    private final List<E> replacedItems;
    private final List<E> newItems;
    private final IntRange indices;

    private ReplaceInList(List<E> replacedItems, List<E> newItems, IntRange indices) {
      this.replacedItems = replacedItems;
      this.newItems = newItems;
      this.indices = indices;
    }

    @Override
    public List<E> replacedItems() {
      return replacedItems;
    }

    @Override
    public List<E> newItems() {
      return newItems;
    }

    @Override
    public IntRange indices() {
      return indices;
    }
  }

  private class ObservableChannels extends GenericObservableChannels<List<E>, MutationEvent>
      implements ObservableList.ObservableChannels<E> {


    protected ObservableChannels() {
      super(
          Flowable.<List<E>>create(
                  flowableEmitter -> {
                    flowableEmitter.onNext(getState());
                    flowableEmitter.onComplete();
                  },
                  BackpressureStrategy.BUFFER)
              .concatWith(mutationEventProcessor.map(MutationEvent::state)),
          Flowable.<MutationEvent>create(
                  flowableEmitter -> {
                    flowableEmitter.onNext(generateMutationEventForNewSubscription());
                    flowableEmitter.onComplete();
                  },
                  BackpressureStrategy.BUFFER)
              .concatWith(mutationEventProcessor));
    }
  }

  private class MutationEvent
      extends GenericMutationEvent<List<E>, List<ObservableList.ListOperation<E>>>
      implements ObservableList.MutationEvent<E> {
    private MutationEvent(List<E> state, List<ListOperation<E>> operations) {
      super(state, operations);
    }
  }

  private MutationEvent generateMutationEventForNewSubscription() {
    List<E> state = getState();
    return new MutationEvent(
        state, ImmutableList.of(new AddToList<>(state, IntRange.startingAt(0).endingAt(state.count()))));
  }
}

File: src/omnia/data/structure/observable/writable/WritableObservableListImpl.kt
package omnia.data.structure.observable.writable

import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.ObservableEmitter
import io.reactivex.rxjava3.subjects.PublishSubject
import io.reactivex.rxjava3.subjects.Subject
import omnia.data.structure.Collection
import omnia.data.structure.IntRange
import omnia.data.structure.List
import omnia.data.structure.immutable.ImmutableList
import omnia.data.structure.observable.ObservableList
import omnia.data.structure.observable.ObservableList.ListOperation
import java.util.OptionalInt
import java.util.function.BiFunction
import java.util.function.Predicate
import java.util.stream.Stream

internal class WritableObservableListImpl<E> : WritableObservableList<E> {
    @Volatile
    private var currentState: ImmutableList<E> = ImmutableList.empty()
    private val mutationEvents: Subject<MutationEvent> = PublishSubject.create<MutationEvent>().toSerialized()
    private val observableChannels: ObservableChannels = ObservableChannels()
    
    override fun insertAt(index: Int, item: E) {
        val range: IntRange = IntRange.just(index)
        mutateState(
                { true },
                { state -> state.toBuilder().insertAt(index, item).build() }
        ) { _, currentState -> generateInsertAtMutations(currentState, range) }
    }

    override fun removeAt(index: Int): E {
        val range: IntRange = IntRange.just(index)
        var value: E
        synchronized(this) {
            value = state.itemAt(index)
            mutateState(
                    { true },
                    { state -> state.toBuilder().removeAt(index).build() }
            ) { previousState: ImmutableList<E>, _: ImmutableList<E> -> generateRemoveAtMutations(previousState, range) }
        }
        return value
    }

    override fun replaceAt(index: Int, item: E): E {
        val range: IntRange = IntRange.just(index)
        var value: E
        synchronized(this) {
            value = state.itemAt(index)
            mutateState(
                    { true },
                    { state -> state.toBuilder().replaceAt(index, item).build() }
            ) { previousState, currentState -> generateReplaceAtMutations(previousState, currentState, range) }
        }
        return value
    }

    override fun add(item: E) {
        mutateState(
                { true },
                { state -> state.toBuilder().add(item).build() }
        ) { _, currentState -> generateInsertAtMutations(currentState, IntRange.just(currentState.count() - 1)) }
    }

    override fun addAll(items: Collection<out E>) {
        mutateState(
                { true },
                { state -> state.toBuilder().addAll(items).build() }
        ) { previousState, currentState ->
            generateInsertAtMutations(
                    currentState,
                    IntRange.startingAt(previousState.count()).endingAt(currentState.count()))
        }
    }

    override fun removeUnknownTyped(item: Any?): Boolean {
        synchronized(this) {
            val index = state.indexOf(item)
            return mutateState(
                    { index.isPresent },
                    { state -> state.toBuilder().removeAt(index.orElseThrow()).build() }
            ) { previousState, _ ->
                generateRemoveAtMutations(
                        previousState, IntRange.just(index.orElseThrow()))
            }
        }
    }

    override fun clear() {
        synchronized(this) {
            mutateState({ obj: ImmutableList<E> -> obj.isPopulated },
                    { ImmutableList.empty() }
            ) { previousState, _ ->
                generateRemoveAtMutations(
                        previousState, IntRange.startingAt(0).endingAt(previousState.count()))
            }
        }
    }

    private fun mutateState(
        shouldMutate: Predicate<ImmutableList<E>>,
        mutator: java.util.function.Function<ImmutableList<E>, ImmutableList<E>>,
        mutationsGenerator: BiFunction<ImmutableList<E>, ImmutableList<E>, List<ListOperation<E>>>
    ): Boolean {
        synchronized(this) {
            val previousState = currentState
            if (!shouldMutate.test(previousState)) {
                return false
            }
            val newState = mutator.apply(previousState)
            currentState = newState
            mutationEvents.onNext(
                    MutationEvent(newState, mutationsGenerator.apply(previousState, newState)))
            return true
        }
    }

    override fun iterator(): Iterator<E> {
        return state.iterator()
    }

    override fun containsUnknownTyped(item: Any?): Boolean {
        return state.containsUnknownTyped(item)
    }

    override fun count(): Int {
        return state.count()
    }

    override val isPopulated: Boolean
        get() = state.isPopulated

    override fun itemAt(index: Int): E {
        return state.itemAt(index)
    }

    override fun indexOf(item: Any?): OptionalInt {
        return state.indexOf(item)
    }

    override fun stream(): Stream<E> {
        return state.stream()
    }

    override fun toReadOnly(): ObservableList<E> {
        return object : ObservableList<E> {
            override fun observe(): ObservableList.ObservableChannels<E> {
                return this@WritableObservableListImpl.observe()
            }

            override fun iterator(): Iterator<E> {
                return this@WritableObservableListImpl.iterator()
            }

            override fun containsUnknownTyped(item: Any?): Boolean {
                return this@WritableObservableListImpl.containsUnknownTyped(item)
            }

            override fun count(): Int {
                return this@WritableObservableListImpl.count()
            }

            override fun itemAt(index: Int): E {
                return this@WritableObservableListImpl.itemAt(index)
            }

            override fun indexOf(item: Any?): OptionalInt {
                return this@WritableObservableListImpl.indexOf(item)
            }

            override fun stream(): Stream<E> {
                return this@WritableObservableListImpl.stream()
            }
        }
    }

    override fun observe(): ObservableChannels {
        return observableChannels
    }

    private val state: ImmutableList<E>
        get() {
            synchronized(this) { return currentState }
        }

    private class AddToList<E>(private val items: List<E>, private val indices: IntRange) : ObservableList.AddToList<E> {
        override fun items(): List<E> {
            return items
        }

        override fun indices(): IntRange {
            return indices
        }
    }

    private class MoveInList<E>(private val items: List<E>, private val previousIndices: IntRange, private val currentIndices: IntRange) : ObservableList.MoveInList<E> {
        override fun items(): List<E> {
            return items
        }

        override fun previousIndices(): IntRange {
            return previousIndices
        }

        override fun currentIndices(): IntRange {
            return currentIndices
        }
    }

    private class RemoveFromList<E>(private val items: List<E>, private val indices: IntRange) : ObservableList.RemoveFromList<E> {
        override fun items(): List<E> {
            return items
        }

        override fun indices(): IntRange {
            return indices
        }
    }

    private class ReplaceInList<E>(private val replacedItems: List<E>, private val newItems: List<E>, private val indices: IntRange) : ObservableList.ReplaceInList<E> {
        override fun replacedItems(): List<E> {
            return replacedItems
        }

        override fun newItems(): List<E> {
            return newItems
        }

        override fun indices(): IntRange {
            return indices
        }
    }

    inner class ObservableChannels : GenericObservableChannels<List<E>, MutationEvent>(
            Observable.create { flowableEmitter: ObservableEmitter<List<E>> ->
                flowableEmitter.onNext(state)
                flowableEmitter.onComplete()
            }
                    .concatWith(mutationEvents.map { obj: MutationEvent -> obj.state() }),
            Observable.create {
                flowableEmitter: ObservableEmitter<MutationEvent> ->
                flowableEmitter.onNext(generateMutationEventForNewSubscription())
                flowableEmitter.onComplete()
            }
                .concatWith(mutationEvents)), ObservableList.ObservableChannels<E>

    inner class MutationEvent(state: List<E>, operations: List<ListOperation<E>>) : GenericMutationEvent<List<E>, List<ListOperation<E>>>(state, operations), ObservableList.MutationEvent<E>

    private fun generateMutationEventForNewSubscription(): MutationEvent {
        val state: List<E> = state
        return MutationEvent(
                state, ImmutableList.of(AddToList(state, IntRange.startingAt(0).endingAt(state.count()))))
    }

    companion object {
        private fun <E> generateInsertAtMutations(
                state: ImmutableList<E>, range: IntRange): ImmutableList<ListOperation<E>> {
            val builder: ImmutableList.Builder<ListOperation<E>> = ImmutableList.builder()

            // Move must be done before insert to make way for new items
            if (range.end() < state.count()) {
                val numItemsMoved = state.count() - range.end()
                val moveFromRange: IntRange = IntRange.startingAt(range.start()).withLength(numItemsMoved)
                val moveToRange: IntRange = IntRange.startingAt(range.end()).withLength(numItemsMoved)
                builder.add(MoveInList(state.getSublist(moveToRange), moveFromRange, moveToRange))
            }
            builder.add(AddToList(state.getSublist(range), range))
            return builder.build()
        }

        private fun <E> generateRemoveAtMutations(
                previousState: ImmutableList<E>, range: IntRange): ImmutableList<ListOperation<E>> {
            val builder: ImmutableList.Builder<ListOperation<E>> = ImmutableList.builder()

            /// Removal must be done BEFORE move to make space for moved items
            builder.add(RemoveFromList(previousState.getSublist(range), range))
            if (range.end() < previousState.count()) {
                val numItemsMoved = previousState.count() - range.end()
                val moveFromRange: IntRange = IntRange.startingAt(range.end()).withLength(numItemsMoved)
                val moveToRange: IntRange = IntRange.startingAt(range.start()).withLength(numItemsMoved)
                builder.add(
                        MoveInList(previousState.getSublist(moveFromRange), moveFromRange, moveToRange))
            }
            return builder.build()
        }

        private fun <E> generateReplaceAtMutations(
                previousState: ImmutableList<E>, currentState: ImmutableList<E>, range: IntRange): ImmutableList<ListOperation<E>> {
            return ImmutableList.of(
                    ReplaceInList(
                            previousState.getSublist(range), currentState.getSublist(range), range))
        }
    }
}